File: matlab/pptx2reveal.m
function pptx2reveal(varargin)
%PPTX2REVEAL converts PPTX into HTML 5 presentation based on the original PPTX and its exportation as high-resolution PNG images
%       pptx2reveal('property1','value1,'property2','value2,...)
%       pptx2reveal(...,'static')
%
%   Example (different from default)
%{
      pptx2reveal('pptx','HDR_defense7d.pptx',...
                  'wwwroot','/home/olivi/Documents/fitness/sandbox/hdr/',...
                  'pptxfolder','/home/olivi/Documents/fitness/sandbox',...
                  'title','HDR May 5th, 2017 *(taken as example)*',...
                  'theme','league',...
                  'static' ...
                 )
% Other example
      pptx2reveal('pptx','Vitrac_packforum2018.pptx',...
                  'wwwroot','/home/olivi/Documents/fitness/sandbox/forum/',...
                  'pptxfolder','/home/olivi/Documents/fitness/sandbox',...
                  'title','Predicting the safety of food contact articles - New science and dicdgital opportunities - October 4, 2018  - Four Points by Sheraton, Kalandergasse 1, Zurich',...
                  'theme','moon',...
                  'static' ...
                 )

%}
% 20/09/2018 - INRA\Olivier Vitrac - rev 09/12/2018

% Revision history
% 20/09/2018 alpha/primitive script (first demo online)
% 22/09/2018 entering in beta version, function, documentation and examples (most of reveal options are implemented)
% 23/09/2018 clean trailing characters ',', ';', ...
% 09/12/2018 add keyword simulation

%% Definitions
vrsn = 0.12;
launchpath = pwd;
cleanupObj = onCleanup(@()cd(launchpath));
switch localname
    case 'LX-Olivier2018'
        local = '/home/olivi/Documents/fitness/sandbox';
    otherwise
        error('define your machine, first')
end
revealoptions_default = struct( ... see https://github.com/hakimel/reveal.js#slide-attributes
    'transition','convex',... none/fade/slide/convex/concave/zoom
    'backgroundTransition','fade',...
    'viewDistance',5,...
    'center',false,...
    'overview',true,...
    'touch',true,...
    'progress',true,...
    'controls',true,...
    'controlsTutorial',true,...
    'controlsLayout','bottom-right',...
    'help',true,...
    'mouseWheel',true,...
    'slideNumber', 'c/t',...
    'autoSlide',15000,...
    'width',"102%",...
	'height',"102%",...
	'margin',0,...
	'minScale',0.1,...
	'maxScale',2,...
    'display','block',...
    'transitionSpeed','default' ... default/fast/slow
    );
default = struct(...
    'wwwroot',fullfile(local,'migration'),...
    'pptxfolder',local,...
    'pptx','Cours_migration2018.pptx',... source pptx required
    'imfolder','jpg',...
    'imsourcefolder','png',...
    'impat','Slide*.jpg',...
    'author','INRA\\Olivier Vitrac',...
    'contact','olivier.vitrac@agroparistech.fr',...
    'title','CONTAMINATION OF FOOD BY PACKAGING MATERIALS',...
    'vrsn',vrsn,...
    'mdfile','index.md',...
    'theme','moon',... % beige,blcak,blood,league,moon,night,serif,simple,sky,solarized,white (https://github.com/hakimel/reveal.js/tree/master/css/theme)
    'initialscale',0.8,...
    'maximumscale',2.0,...
    'userscalable','yes',...
    'reveal',revealoptions_default);
boolean = {'false','true'};
keyword = {'static','simulation' 'repair'};
staticobjects = {
    'css' 'js'   'lib' 'plugin' 'index.html'
    'dir' 'dir'  'dir' 'dir'    'file'
}; % content of the static objects

%% arg check
[revealoptions,others] = argcheck(varargin,struct('reveal',revealoptions_default),keyword,'keep','case','nostructexpand');
o = argcheck(others,rmfield(default,'reveal'),'','case');
o.static = revealoptions.static;
o.simulation = revealoptions.simulation;
o.repair = revealoptions.repair;
revealoptions = argcheck(revealoptions.reveal,revealoptions_default,'','case');

if ~exist(o.wwwroot,'dir'), error('the wwwroot folder ''%s'' does not exist',o.wwwroot), end
if ~exist(o.pptxfolder,'dir'), error('the pptx folder ''%s'' does not exist',o.pptxfolder), end
if ~exist(fullfile(o.pptxfolder,o.pptx),'file'), error('the PPTX file ''%s'' does not exist in ''%s''',o.pptx,o.pptxfolder), end

imenginecmd = sprintf('mogrify -path ''%s'' -resize 25x25%% -quality 90 -format jpg *.PNG',fullfile(o.wwwroot,o.imfolder));
imengine = @() system(imenginecmd);
textenginecmd = @(i) sprintf('unzip -qc ''%s'' ppt/slides/slide%d.xml | grep -oP ''(?<=\\<a:t\\>).*?(?=\\</a:t\\>)''',fullfile(o.pptxfolder,o.pptx),i);
textengine = @(i) system(textenginecmd(i));
slide = @(i) getOutput(textengine,2,i);

%% static output
[~,fnpptx] = fileparts(o.pptx);
staticfolder = fullfile(o.pptxfolder,sprintf('_html_%s',fnpptx));
existstaticfolder = exist(staticfolder,'dir');

%% search for thumb images
if ~exist(fullfile(o.wwwroot,o.imfolder),'dir')
    dispf('PPTX2REVEAL: the image folder ''%s'' does not exist in ''%s''',o.imfolder,o.wwwroot)
    if exist(fullfile(o.wwwroot,o.imsourcefolder),'dir')
        cd(fullfile(o.wwwroot,o.imsourcefolder))
        dispf('PPTX2REVEAL: conversion of all images from the folder ''%s''',o.imsourcefolder)
        if o.simulation,    dispf('simulation>> mkdir %s',fullfile(o.wwwroot,o.imfolder))
        else,               mkdir(fullfile(o.wwwroot,o.imfolder))
        end
        cd(fullfile(o.wwwroot,o.imsourcefolder))
        if o.simulation,    dispf('simulation>> %s',imenginecmd)            
        else,               dispf('run>> %s',imenginecmd), imengine();
        end
        cd(launchpath)
    else
        error('the image source folder ''%s'' does not exist in ''%s''',o.imsourcefolder,o.wwwroot)
    end
end
f = explore(o.impat,fullfile(o.wwwroot,o.imfolder),[],'abbreviate');
if isempty(f)
    if o.simulation
        dispf('simulation>> ERROR no files ''%s'' found in ''%s''',o.impat,fullfile(o.wwwroot,o.imfolder))
        f = explore('*.PNG',fullfile(o.wwwroot,'png'),[],'abbreviate');
        if isempty(f)
            error('SIMULATION failed: no jpg or png files, please export slides first')
        end
    else
        error('no files ''%s'' found in ''%s''',o.impat,fullfile(o.wwwroot,o.imfolder))
    end
end
nn = str2double(uncell(regexp({f.name},'(\d+$)','tokens')));
[~,ind] = sort(nn);
f = f(ind); n = length(f);
dispf('PPTX2REVEAL: %d slides found',n)

%% generate the MarkDown File
slidetransition = '<!--s-->'; %'\n---\n';
verticalbar = '<!--v-->';
if o.simulation
    dispf('simulation>> touch %s',fullfile(o.wwwroot,o.mdfile))
    disp([repmat('-',1,80) '<<< begin'])
    fid = 1;
else
    if ~exist(fullfile(o.wwwroot,o.mdfile),'file') || o.repair
        fid = fopen(fullfile(o.wwwroot,o.mdfile),'w');
    end
end
fprintf(fid,'---\n');
fprintf(fid,'title: %s\n',o.title);
fprintf(fid,'separator: %s\n',slidetransition);
fprintf(fid,'verticalSeparator: %s\n',verticalbar);
fprintf(fid,'theme: %s\n',o.theme);
fprintf(fid,'revealOptions:\n');
for p=fieldnames(revealoptions)'
    if ~isempty(revealoptions.(p{1}))
        if isnumeric(revealoptions.(p{1}))
            fprintf(fid,'    %s: %d\n',p{1},revealoptions.(p{1}));
        elseif islogical(revealoptions.(p{1}))
            fprintf(fid,'    %s: %s\n',p{1},boolean{1+revealoptions.(p{1})});
        else
            fprintf(fid,'    %s: ''%s''\n',p{1},revealoptions.(p{1}));
        end
    end
end
fprintf(fid,'---\n');
fprintf(fid,'___\n');
fprintf(fid,'## %s\n',o.title);
fprintf(fid,'### %s\n',o.author);
fprintf(fid,'___\n');
fprintf(fid,'    version: demo (%0.3g)\n',o.vrsn);
fprintf(fid,'    date: %s\n',datestr(now,'mmmm dd, yyyy'));
fprintf(fid,'    content: %d slides\n',n);
fprintf(fid,'    source: %s\n',o.pptx);
fprintf(fid,'    contact: %s\n',o.contact);
fprintf(fid,slidetransition);
screen = '';
for i=1:n
    screen = dispb(screen,'>> converting slide %d of %d',i,n);
    content = slide(i);
    if isempty(content), lines = {''}; else, lines = deblank(strsplit(content,newline)); end
    if length(lines{1})<4 || any(lines{1}<32)
        fprintf(fid,'\n### slide %d/%d \n',i,n);
    else
        if strcmpi(lines{1}(end-1:end),'of') || strcmpi(lines{1}(end-2:end),'and')
            slidetxt = [lines{1} ' ' lines{2}];
        else
            slidetxt = lines{1};
        end
        slidetxt = regexprep(slidetxt,'\s*|\,|\;|\.|\+|\-|\_','');
        fprintf(fid,'\n### %s \n',slidetxt);
    end
    fprintf(fid,'![%s](./%s/%s  "slide %d of %d") \n',f(i).file,o.imfolder,f(i).file,i,n);
    if ~isempty(content), fprintf(fid,'note: %s',content); end
    fprintf(fid,slidetransition);
end
if o.simulation,         disp([repmat('-',1,80) '<<< end'])
else,                    fclose(fid);
end

%% display
if o.static
    cd(o.wwwroot)
    [~,fname] = fileparts(o.pptx);
    staticsite = sprintf('_html_%s',fname);
    revealcmd = sprintf('reveal-md %s --static ''%s''',o.mdfile,staticsite);
    if o.simulation
        dispf('simulation>> %s',revealcmd)
    else
        if o.repair
            if existstaticfolder, dispf('clean>> %s',staticfolder), rmdir(staticfolder,'s'), end
            for sobj = staticobjects
                    if exist(fullfile(o.pptxfolder,sobj{1}),sobj{2})
                        dispf('clean>> %s',fullfile(o.pptxfolder,sobj{1}))
                        if strcmp(sobj{2},'dir'), rmdir(fullfile(o.pptxfolder,sobj{1}),'s')
                        else,                     delete(fullfile(o.pptxfolder,sobj{1}))
                        end % dir, file
                    end % if exist
            end % next sobj
        end % if repair
        dispf('run>> %s',revealcmd), system(revealcmd);
        if ~exist(staticfolder,'dir'), error('the following command fails: %s',revealcmd), end
    end % simulation
    idxfile = fullfile(o.wwwroot,staticsite,'index.html');
    if o.simulation
        dispf('simulation>> update %s',idxfile)
    else
        % edit index.html
        fid = fopen(idxfile,'r'); raw=uncell(textscan(fid,'%s','delimiter','\n','whitespace','')); fclose(fid);
        iviewport = ~cellfun(@isempty,regexp(raw,'<meta name="viewport".*>'));
        raw{iviewport} = sprintf('<meta name="viewport" content="width=device-width, initial-scale=%0.1f, maximum-scale=%0.1f, user-scalable=%s">',...
            o.initialscale,o.maximumscale,o.userscalable);
        fid = fopen(idxfile,'w'); for i=1:length(raw); fprintf(fid,'%s\n',raw{i}); end, fclose(fid);
        % move content (only if repair, if not do it manually)
        if o.repair
            for sobj = staticobjects
                if exist(fullfile(staticfolder,sobj{1}),sobj{2})
                dispf('move>> %s',fullfile(o.pptxfolder,sobj{1}))
                movefile(fullfile(staticfolder,sobj{1}),fullfile(o.pptxfolder,sobj{1}))
                end % if exist
            end % next sobj
            dispf('delete>> %s',staticfolder), rmdir(staticfolder)
        end % if repair
    end % simulation
else
    cd(o.wwwroot)
    dispf('\nOnline demo open in the default browser')
    revealcmd = sprintf('reveal-md %s',o.mdfile);
    if o.simulation,    dispf('simulation>> %s',revealcmd)   
    else,               dispf('run>> %s',revealcmd), system(revealcmd);
    end
end


end % end function

% ===================================== PRIVATE FUNCTIONS =====================================
function varargout = getOutput(func,outputNo,varargin)
    varargout = cell(max(outputNo),1);
    [varargout{:}] = func(varargin{:});
    varargout = varargout(outputNo);
end

%% __NOTES__
% Image resize
%   mogrify -path /home/olivi/Documents/sandbox/migration/jpg -resize 25x25% -quality 90 -format jpg *.PNG
% Text extraction
%   unzip -qc Cours_migration2018.pptx ppt/slides/slide1.xml | grep -oP '(?<=\<a:t\>).*?(?=\</a:t\>)'
%   unzip Cours_migration2018.pptx 'ppt/slides/*.xml' -d .  
% HMTL5 conversion
% reveal-md migration.md --static
% reveal-md slides.md --static _site

File: fitness/pptx2reveal.py
# -*- coding: utf-8 -*-
"""
Created on Tue Apr  5 11:05:56 2022

@author: olivi
"""

# %% Dependencies
import os, fnmatch, socket, subprocess, re
from pathlib import PurePosixPath, PureWindowsPath
from datetime import datetime
import numpy as np
from fitness.private.struct import struct


# %% class definitions
class config(struct):
    """ mypptx class """
    _type = "reveal"        # object type
    _fulltype = "reveal-md" # full name
    _ftype = "propertie"    # field name
    _maxdisplay = 60

class filedetails(struct):
    """ explore class, almost as the original explore with abbreviate """
    _type = "file"        # object type
    _fulltype = "file details" # full name
    _ftype = "attribute"    # field name
    _maxdisplay = 120
    
    def __init__(self,file):
        """ initialization based on a valid filename """
        super().__init__(fullfilename=file)
        if not os.path.isfile(file):
            raise IOError(f'the file "{file}" does not exist')
        self.path, self.filename = os.path.split(file)
        self.name, self.ext = os.path.splitext(self.filename)
        self.date = os.path.getmtime(file)
        self.size = os.path.getsize(file)
        self.datestr = datetime.fromtimestamp(self.date).strftime("%Y-%m-%d %H:%M:%S")
        
    
# %% Functions mimicking the ones in version 1, which were initially coded in Matlab
def mem2bytes(t):
    """
        MEM2BYTES converts memory size (from str) to bytes (to double)
        v = mem2byes("300 kB")

    """
    if not isinstance(t,str): raise ValueError("t must be a string")
    U = "KMGTP";
    f = [1<<10,1<<20,1<<30,1<<40,1<<50] # A left shift by n bits is defined as multiplication with pow(2,n)
    v0 = re.search('(?P<v>\d+\.?\d*)\s*(?P<u>[KMGTPkmgtp]?)[bB]',t)
    if v0 is None: raise ValueError(f'unable to convert in bytes the string "{t}"')
    v = float(v0.group("v"))
    u = v0.group("u")
    if u !="": v = v*f[U.find(u.upper())]
    return v

def explore(filePattern="Slide*.png",directory=os.getcwd(),topdown=True,onerror=None):
    """
        EXPLORE lists all files matching a pattern in a given directory (unlimited depth)
        files = explore("*.PNG",".")
    """
    matches = []
    for path, dirs, files in os.walk(os.path.abspath(directory),topdown=topdown,onerror=onerror):
        for filename in fnmatch.filter(files, filePattern):
            matches.append(filedetails(os.path.join(path, filename)))
    return matches

def system(cmd):
    """ system() equivalent to Matlab one
        cmd = (code,params)
        returns 0 if no error as well as the standard output
        note: os.system() is depreciated and does not capture the output
    """
    out = subprocess.run(cmd,capture_output=True, text=True)
    return out.returncode, out.stdout.strip("\n")
    
# %% default configurations
vrsn = 0.331;
launchpath = os.getcwd();
onCleanup = lambda : os.chdir(launchpath)
hostname = socket.gethostname()
if hostname == "LX-Olivier2020":
    local = "/home/olivi/Documents/fitness/sandbox"
elif hostname=="lne-ThinkPad-P52":
    local = '/home/mai/Documents/fitness/sandbox'
else:
    raise NameError('define your machine, first')

defaultrevealoptions = config(
        # see https://github.com/hakimel/reveal.js#slide-attributes
        transition = "convex", # none/fade/slide/convex/concave/zoom
        backgroundTransition = "fade",
        viewDistance = 5,
        center = False,
        overview = True,
        touch = True,
        progress = True,
        controls = True,
        controlsTutorial = True,
        controlsLayout = "bottom-right",
        help = True,
        mouseWheel = True,
        slideNumber = "c/t",
        autoSlide = 15000,
        width = "102%",
    	height =   "102%",
    	margin = 0,
    	minScale = 0.1,
    	maxScale = 3,
        display = "block",
        transitionSpeed = "default" # default/fast/slow   
    )

dependencies = ["references","videos","movies","extra","casestudies","guidelines","howto","solutions"]

defaultTOC = config();
defaultreplace = config();
defaultsubstitute = config();
boolean = ['false','true'];
keyword = ['static','simulation' 'repair' 'forcedupdate'];
staticobjects = {
    "css":"dir",
    "js":"dir",
    "lib":"dir",
    "plugin":"dir",
    "index.html":"file"
    } # content of the static objects
imsourcefolder_default = 'png';

# %% main()
def pptx2reveal(
    wwwroot = local,
    wwwrel = "demo",
    wwwlib = "src",
    pptxfolder = local,
    pptx = "Cours_migration2018.pptx", # source pptx required
    imfolder = "jpg",
    imsourcefolder = '',
    impat = "Slide*.jpg" ,  # assets (raster)
    svgpat = "Slide*.svg",  # assets (svg)
    author = "INRA\\Olivier Vitrac",
    contact = "olivier.vitrac@agroparistech.fr",
    YAMLtitle = "", # title will be used when empty
    title = "CONTAMINATION OF FOOD BY PACKAGING MATERIALS",
    vrsn = vrsn,
    mdfile = "index.md",
    projectname = "",
    theme = "moon", #beige,black,blood,league,moon,night,serif,simple,sky,solarized,white (https://github.com/hakimel/reveal.js/tree/master/css/theme)
    resize = "25x25%",
    initialscale = 0.8,
    maximumscale = 2.0,
    userscalable = "yes",
    revealoptions = None,
    TOC = None,
    replace = None,
    substitute = None,
    headingshift = "", # add '##' to convert h1->h3, h2->h4, h3->h5, etc.
    headingshiftnotes = "##",
    fontsize = np.array([0.6000,0.5500,0.4500,0.3500,0.3000])+.1, # h1 h2 h3, etc. in em (to be used in css/theme/theme.css)
    hash = "",
    traceability = True,
    maxtitle = 120,# max length of title
    maxcontentslide = 12, #maximum number of lines on the content slide
    favicon = "~/Bulk/fitness2019/template/favicon/favicon.ico",
    mainindex = "",
    dependencies = dependencies,
    dependencymain = "index.html",
    imformatpreferred = "png",
    svgmaxsize = "2 MB"
        ):
    
    # -- arg check -- (try to mimic the original Matlab code)
    if revealoptions is None: revealoptions = defaultrevealoptions
    elif isinstance(revealoptions,struct): revealoptions.check(defaultrevealoptions)
    else: raise TypeError("revealoptions should be a struct() or a config() object")
    if TOC is None: TOC = defaultTOC
    elif isinstance(TOC,struct): TOC.check(defaultTOC)
    else: raise TypeError("TOC should be a struct() or a config()")
    if replace is None: TOC = defaultreplace
    elif isinstance(replace,struct): replace.check(defaultreplace)
    else: raise TypeError("replace should be a struct() or a conf() object")
    if substitute is None: substitute = defaultsubstitute
    elif isinstance(substitute,struct): substitute.check(defaultsubstitute)
    else: raise TypeError("substitute should be a struct() or a conf() object")                      
    static = revealoptions.static
    simulation = revealoptions.simulation
    repair = revealoptions.repair
    forcedupdate = revealoptions.forcedupdate
    if projectname == "":
        _,projectname = os.path.split(mdfile)
        projectname,_ = os.path.splitext(projectname)
    if YAMLtitle == "": YAMLtitle = title
    outputhtml = projectname + ".html"
    staticobjects[outputhtml] = "file"


    # -- folder check --
    # wwwroot = server root
    # www = /wwwroot/.../wwwrel/ = location of index.html
    # wwwlib = /wwwroot/.../o.wwwlib/ = location of libraries
    if not os.path.isdir(wwwroot): raise IOError('the wwwroot folder "%s" does not exist' % wwwroot)
    if not os.path.isdir(pptxfolder): raise IOError('the pptx folder "%s" does not exist' % pptxfolder)
    if not os.path.isfile(os.path.join(pptxfolder,pptx)): raise IOError('the PPTX file ''%s'' does not exist in "%s"' % (pptx,pptxfolder))
    www = os.path.join(wwwroot,wwwrel)  # location of index.html
    wwwlib = os.path.join(wwwroot,wwwlib)
    if not os.path.isdir(www): os.mkdir(www)
    if not os.path.isdir(wwwlib): os.mkdir(wwwlib)
    # htmlpath(p) make p relative to root
    # -- example --
    # tmpwww = 'www/folder1/html'
    # tmpwwwlib = 'lib'
    # htmloldlibpath('js') -> './js'
    # htmllibpath('js') ->  './../../../lib/js'
    #
    # // Matlab code //
    # pathsplit = @(p) strsplit(regexprep(p,{['\' filesep '+'],['^\' filesep],['\' filesep '$']},{['\' filesep] '' ''}),filesep);
    # fullfilex = @(x) fullfile(x{:});
    # backpath = @(p) fullfilex(repmat({'..'},length(pathsplit(p)),1));
    # htmlpath = @(p) fullfile('./',backpath(tmpwww),p);
    # htmloldlibpath = @(p) fullfile('./',fullfilex(pathsplit(p)));
    # htmllibpath = @(p) fullfile(htmlpath(tmpwwwlib),p);
    # -------------
    wwwroot = PureWindowsPath(wwwroot).as_posix()
    tmpwww = PureWindowsPath(wwwrel).as_posix() 
    tmpwwwlib = PureWindowsPath(wwwlib).as_posix()
    safepath = lambda p: PureWindowsPath(p).as_posix()
    backpath = lambda p: PurePosixPath(safepath(os.path.relpath(".",p)))
    htmlpath = lambda p: PurePosixPath("./")/backpath(tmpwww)/safepath(p)
    htmloldlibpath = lambda p: PurePosixPath("./")/safepath(p)
    htmllibpath = lambda p: PurePosixPath(tmpwwwlib)/safepath(p)
    
    # -- engines -- (requires mogrify, libreoffice and pdf2svg)
    #imenginecmd = sprintf('mogrify -path ''%s'' -resize 25x25%% -quality 90 -format jpg *.PNG',fullfile(www,o.imfolder));
    _,fname = os.path.split(pptx)
    fname,_ = os.path.splitext(fname)
    pdf = "%s.pdf" % fname
    imenginecmd  = ("mogrify" "-path '%s' -resize %s -quality 90 -format jpg *.PNG" % \
        (os.path.join(www,imfolder),resize) );
    pdfenginecmd = ("libreoffice", "--headless --convert-to pdf '%s'" % \
        os.path.join(pptxfolder,pptx) )
    svgenginecmd = ("pdf2svg","'%s' '%s' all" % \
        ( os.path.join(pptxfolder,pdf), os.path.join(www,imfolder,'Slide%d.svg') ) )
    
    imengine  = lambda: system(imenginecmd);
    pdfengine = lambda: system(pdfenginecmd);
    svgengine = lambda: system(svgenginecmd);
    svgmaxsize = mem2bytes(svgmaxsize);
    svgpreferred = imformatpreferred.lower()=="svg"
    
    # -- static output --
    # location:: www = /wwwroot/.../wwwrel/
    relpptxfolder,fnpptx = os.path.split(pptx)
    fnpptx,_ = os.path.splitext(fnpptx)
    if projectname == "": 
        staticsite = "_html_%s" % fnpptx
    else:
        staticsite = "_html_%s" % projectname    
    staticfolder = os.path.join(www,staticsite)
    existstaticfolder = os.path.isdir(staticfolder)
    
    
    # -- search for thumb images --
    # location:: www = /wwwroot/.../wwwrel/imfolder/
    if not os.path.isdir(os.path.join(www,imfolder)):
        print('PPTX2REVEAL: the image folder "%s" does not exist in "%s"' % (imfolder,wwwroot))
        if imsourcefolder == "":
            if os.path.isdir(os.path.join(pptxfolder,relpptxfolder,fnpptx)):
                rootimsourcefolder = os.path.join(pptxfolder,relpptxfolder)
                imsourcefolder = fnpptx
            else:
                rootimsourcefolder = wwwroot
                imsourcefolder = imsourcefolder_default
        else:
                rootimsourcefolder = wwwroot
        if os.path.isdir(os.path.join(rootimsourcefolder,imsourcefolder)):
            print('PPTX2REVEAL: conversion of all images from the folder "%s"' % imsourcefolder)
            if simulation: print('simulation>> mkdir %s' % os.join(www,imfolder))
            else: os.mkdir(os.join(www,imfolder))
            os.chdir(os.join(rootimsourcefolder,imsourcefolder))
            if simulation: print('simulation>> cd ''%s''\nsimulation>> %s' % (os.path.join(rootimsourcefolder,imsourcefolder),imenginecmd))
            else: print('run>> %s' % imenginecmd); imengine();
            os.chdir(launchpath)
        else:
            raise IOError('the image source folder "%s" does not exist in "%s"' % \
                          (imsourcefolder,rootimsourcefolder) )


    f = explore(impat,os.path.join(www,imfolder))
    if svgpreferred:
        if not os.path.isfile(os.path.join(pptxfolder,pdf)):
            print('run>> %s' % pdfenginecmd)
            currentpath = os.getcwd()
            os.chdir(pptxfolder)
            pdfengine()
            os.chdir(currentpath)
        if not os.path.isfile(os.path.join(pptxfolder,pdf)):
            raise IOError('the conversion of "%s" to PDF fails in "%s"' % (pdf,pptxfolder))
        LD_LIBRARY_PATH = os.getenv('LD_LIBRARY_PATH')    # save library path
        os.environ["LD_LIBRARY_PATH"] = ""  # clear library path
        print("run>> %s" % svgenginecmd); svgengine()
        os.environ["LD_LIBRARY_PATH"] = LD_LIBRARY_PATH  # restore the library path        
        fsvg = explore(svgpat,os.path.join(www,imfolder));
    else:
        fsvg = []
    if len(f)==0:
        if simulation:
            print('simulation>> ERROR no files "%s" found in "%s"'  % \
                  (impat,os.path.join(www,imfolder)) )
            f = explore(impat,os.path.join(rootimsourcefolder,imsourcefolder))
            if len(f)==0:
                raise IOError('SIMULATION failed: no image files matching "%s", please export slides first' % impat)
        else:
            raise IOError('no file matching "%s" in "%s"',impat,os.path.join(www,imfolder))
    

    searchslidenum = re.compile('(\d+$)')
    nn = [int(searchslidenum.findall(f0.name)[0]) for f0 in f]
    f = [fimg for _, fimg in sorted(zip(nn, f))]
    n = len(f)    

    if svgpreferred:
        if len(fsvg)==0:
            raise IOError('no file matching "%s" in "%s"' % (svgpat,os.path.join(www,imfolder)))
        nnsvg =  [int(searchslidenum.findall(f0.name)[0]) for f0 in fsvg]
        fsvg =  [fimg for _, fimg in sorted(zip(nnsvg, fsvg))]
        nsvg = len(fsvg)
        if nsvg != n:
            raise IOError(f"the number of raster images ({nn}) does not match the number of vectorial images ({nsvg})")

